vulnerabilities/importers/nvd.py

```diff
@@ -20,150 +20,175 @@
-                return True
-
-        return False
-
-    @staticmethod
-    def extract_cpes(cve_item):
-        cpes = set()
-        for node in cve_item["configurations"]["nodes"]:
-            for cpe_data in node.get("cpe_match", []):
+        )
+        if "https://nvd.nist.gov/vuln/detail/{cve_id}" in ref_urls:
+            ref_urls.remove(f"https://nvd.nist.gov/vuln/detail/{cve_id}")
+        references.extend([Reference(url=url) for url in ref_urls])
+        if cve_id:
+            aliases.append(cve_id)
+        summary = extract_summary(cve_item)
+        yield AdvisoryData(
+            aliases=aliases,
+            summary=summary,
+            references=sorted(references),
+            date_published=dateparser.parse(cve_item.get("publishedDate")),
+        )
+
+
+def extract_reference_urls(cve_item):
+    """
+    Return a list of reference URLs for a given CVE item.
+    """
+    urls = set()
+    for reference in get_item(cve_item, "cve", "references", "reference_data") or []:
+        ref_url = reference.get("url")
+
+        if not ref_url:
+            continue
+
+        if ref_url.startswith("http") or ref_url.startswith("ftp"):
+            urls.add(ref_url)
+
+    return urls
+
+
+def related_to_hardware(cpes):
+    """
+    Return True if the CVE item is related to hardware.
+    """
+    for cpe in cpes:
+        cpe_comps = cpe.split(":")
+        # CPE follow the format cpe:cpe_version:product_type:vendor:product
+        if len(cpe_comps) > 2 and cpe_comps[2] == "h":
+            return True
+
+    return False
+
+
+def extract_cpes(cve_item):
+    """
+    Return a list of CPEs for a given CVE item.
+    """
+    cpes = set()
+    for node in get_item(cve_item, "configurations", "nodes") or []:
+        for cpe_data in node.get("cpe_match") or []:
+            if cpe_data.get("cpe23Uri"):
                 cpes.add(cpe_data["cpe23Uri"])
-        return cpes
+    return cpes
+
+
+def extract_severity_scores(cve_item):
+    """
+    Yield a vulnerability severity for each `cve_item`.
+    """
+    if get_item(cve_item, "impact", "baseMetricV3"):
+        yield VulnerabilitySeverity(
+            system=SCORING_SYSTEMS["cvssv3"],
+            value=str(get_item(cve_item, "impact", "baseMetricV3", "cvssV3", "baseScore")),
+        )
+        yield VulnerabilitySeverity(
+            system=SCORING_SYSTEMS["cvssv3_vector"],
+            value=str(get_item(cve_item, "impact", "baseMetricV3", "cvssV3", "vectorString")),
+        )
+
+    if get_item(cve_item, "impact", "baseMetricV2"):
+        yield VulnerabilitySeverity(
+            system=SCORING_SYSTEMS["cvssv2"],
+            value=str(get_item(cve_item, "impact", "baseMetricV2", "cvssV2", "baseScore")),
+        )
+        yield VulnerabilitySeverity(
+            system=SCORING_SYSTEMS["cvssv2_vector"],
+            value=str(get_item(cve_item, "impact", "baseMetricV2", "cvssV2", "vectorString")),
+        )
+
+
+class NVDBasicImprover(Improver):
```

Comment:
Can you add a docstring that explains what the purpose of this improver is?

Reply:
#681 see this for follow up